packages/vm/src/context.rs

```diff
@@ -135,13 +166,54 @@ pub(crate) fn move_into_context<S: Storage, Q: Querier>(target: &mut Ctx, storag
     b.querier = Some(querier);
 }
 
+pub fn get_gas_state<S: Storage, Q: Querier>(ctx: &mut Ctx) -> &mut GasState {
+    &mut get_context_data::<S, Q>(ctx).gas_state
+}
+
+#[cfg(feature = "default-singlepass")]
+pub fn try_consume_gas<S: Storage, Q: Querier>(ctx: &mut Ctx, used_gas: u64) -> VmResult<()> {
+    use crate::backends::{get_gas_left, set_gas_limit};
+    use crate::VmError;
+
+    let ctx_data = get_context_data::<S, Q>(ctx);
+    if let Some(mut instance_ptr) = ctx_data.wasmer_instance {
+        let instance = unsafe { instance_ptr.as_mut() };
+        let gas_state = &mut ctx_data.gas_state;
+
+        let wasmer_used_gas = gas_state.gas_limit.saturating_sub(get_gas_left(instance));
```

Comment:
All you said was wrong, unfortunately :sweat_smile: 
Yes, internally there's a bit of complication because currently wasmer can only have so much available gas to begin with, but this algorithm is agnostic to that.

---

`GasState` tracks two things: (1)The initial total gas limit for the computation, and (2) the gas that was consumed outside of wasmer.
The `set_gas_limit` function that takes a wasmer instance conceptually resets its used gas to `0`, and sets a new amount of gas units that it can consume from that point on.
So, every time we report gas being consumed externally (that is, `try_consume_gas` is called) we:
1. check how much gas was already consumed in wasmer since the computation started (by subtracting the amount of gas it thinks it has left from the total gas limit)
2. add the newly externally consumed gas to the sum of externally consumed gas.
3. set the new gas limit in wasmer to the amount of gas units left for the calculation (by subtracting the amounts we establish in steps 1. and 2. from the total gas limit).
4. check if the amounts from 1. and 2. together exceed the gas limit.

Now that i look at it again, we can flip steps 3 and 4 as a small optimisation, so we don't even update the wasmer instance if we already know that we ran out of gas.